### PR TITLE
Added pandas and JSON normalization code

### PR DESCRIPTION
**Pull request by IamEzio**
Fixes part of #2895 

This PR does the following:
- Adds `pandas` as a dependency and uses `pandas.json_normalize()` method to normalize JSON file inputs.
- Adds custom exceptions for JSON import feature.
- Adds `missing_keys.json` file to test json normalization code.

**Screenshots**

NA

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by rajatvijay** [comment]
@IamEzio assigning this back to you since python test cases are failing.

**Comment by IamEzio** [comment]
Thanks @rajatvijay, I have fixed them now. PTAL

File: db/records/operations/insert.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import json
+import pandas as pd
```
Comment (kgodey) [inline db/records/operations/insert.py:2]: Why are we renaming the import here?

File: mathesar/imports/json.py [hunk 2]
```diff
@@ -1,24 +1,47 @@
...
 from mathesar.state import reset_reflection
 
 
+def validate_json_format(data_file_content):
+    try:
+        data = json.load(data_file_content)
+    except (JSONDecodeError, ValueError) as e:
+        raise database_api_exceptions.InvalidJSONFormat(e)
+
+    if isinstance(data, list) and all(isinstance(val, dict) for val in data):
```
Comment (dmos62) [inline mathesar/imports/json.py:25]: A variable name to communicate intent of this boolean composition would be nice.

```suggestion
    is_list_of_dicts = isinstance(data, list) and all(isinstance(val, dict) for val in data)
    if is_list_of_dicts:
```

File: db/records/operations/insert.py [hunk 3]
```diff
@@ -32,9 +33,14 @@ def insert_record_or_records(table, engine, record_data):
     return None
 
 
-def insert_records_from_json(table, engine, json_filepath):
+def insert_records_from_json(table, engine, json_filepath, column_names):
```
Comment (dmos62) [inline db/records/operations/insert.py:36]: Add a docstring. Include explanation for why `column_names` is necessary. Include explanation for why and how `json_normalize` is being invoked (what's `max_level` and why is it `0`). Include explanation of the general algorithm and intent.

File: db/records/operations/insert.py (hunk 1 above)
Comment (dmos62) [reply db/records/operations/insert.py:2]: I'm used to seeing `pandas` renamed to `pd`, but most people's readability would probably be improved without this shorthand.

**Review by dmos62 — CHANGES_REQUESTED** [review]
Also, please provide tests for JSON validation and normalization.

File: db/records/operations/insert.py (hunk 1 above)
Comment (kgodey) [reply db/records/operations/insert.py:2]: Agreed re: readability.

Comment (IamEzio) [reply db/records/operations/insert.py]: Updated it now. Thanks!

File: mathesar/imports/json.py (hunk 2 above)
Comment (IamEzio) [reply mathesar/imports/json.py]: Thanks Dom, added it now.

File: db/records/operations/insert.py (hunk 3 above)
Comment (IamEzio) [reply db/records/operations/insert.py:36]: Done. Thanks!

**Comment by IamEzio** [comment]
Hi @kgodey @dmos62 I have made the suggested changes. I also have the PR #2977 as a follow-through that has the required API tests.

File: db/records/operations/insert.py (hunk 3 above)
Comment (dmos62) [reply db/records/operations/insert.py:36]: Include a summary and explation of the algorithm. You normalize into a dataframe, then convert to a JSON string, then to Python, ending up with a sequence of rows where each row is dict-like, in which every key-value pair is a column, then you take those columns and if they are a dict or a list, you serialize them back into JSON. That takes some time to figure out and the reason for having to do some of these things is not necessarily immediately obvious. Help the reader out with a summary and an explanation.

**Review by dmos62 — CHANGES_REQUESTED** [review]

File: db/records/operations/insert.py (hunk 3 above)
Comment (dmos62) [reply db/records/operations/insert.py:36]: Presume that the reader is not me or anyone that you've discussed JSON importing with.

Comment (IamEzio) [reply db/records/operations/insert.py:36]: Hi Dom, I have added the explanation now. PTAL. Thanks!

Comment (dmos62) [reply db/records/operations/insert.py:36]: Awesome!

**Review by dmos62 — APPROVED** [review]

**Pull request merged**